Github action to periodically `cargo update` to keep dependencies current

pitaj opened the pull request:
Opens a PR periodically with the results of `cargo update`. If an unmerged PR for the branch `cargo_update` already exists, it will edit then reopen it if necessary.

~~This also uses [`cargo-upgrades`](https://gitlab.com/kornelski/cargo-upgrades) to provide a list of available major upgrades in the PR body.~~

It includes the list of changes output by `cargo update` in the commit message and PR body. Note that this output is currently sub-optimal due to https://github.com/rust-lang/cargo/issues/9408, but if updates are made more regularly that is less likely to show up.

Example PR: https://github.com/pitaj/rust/pull/2
Example action run: https://github.com/pitaj/rust/actions/runs/5035731903
Prior discussion: https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/dependabot.20updates.3F

Up for discussion:
- What period do we want? Currently weekly
- What user should it use? Currently "Github Actions"
- Do we need the extra security of provided by executing `cargo update` and `cargo-upgrades` in a separate job?
  If not I can simplify it to not need artifacts.
- PR message wording
- PR should probably always be `rollup=always`?
- What branch should it use?
- What should it do if no updates are available? Currently fails the job on empty commit
- Should the yml file live in `src/ci` instead of directly under workflows?
- ~~Is using the latest nightly toolchain enough to ensure compatibility with `Cargo.lock` and `Cargo.toml`s in master?~~
  Now pulls the bootstrap version from stage0.json

r? infra

pitaj committed:
run `cargo update` weekly

- Keep Cargo.lock dependencies current
- Presents output from `cargo update` in commit and PR
- Edit existing open PR, otherwise open a new one
- Skip if existing open PR is S-waiting-on-bors

pitaj commented:
Thanks for taking a look.

> Can we avoid creating the branch in this repository?

I'm not sure if that's possible, but I will look into it. Are you thinking rust-lang-ci/rust instead?

> Can we filter out or skip listing any submodule/subtree dependencies?

Certainly possible, maybe not the simplest thing to do. 

> Following from the last point, I'm not sure how useful the list of pending major version upgrades is. If someone has bandwidth to work on those, it seems like they could also run the command themselves.

Yeah that's fair. I thought it would be useful to see them periodically, but maybe this should be handled separately. Open an issue for each major upgrade instead. Marking them as E-easy would invite community contributions.

> I initially thought that was the delta list, which might've been more helpful

Good idea. I'll look into parsing Cargo.lock and including the set of changes in the commit message and PR body.

Mark-Simulacrum commented:
> I'm not sure if that's possible, but I will look into it. Are you thinking rust-lang-ci/rust instead?

I think that would work.

> Good idea. I'll look into parsing Cargo.lock and including the set of changes in the commit message and PR body.

I think the cargo update command should print this as it runs? Can we just use that?

pitaj commented:
> I think the cargo update command should print this as it runs? Can we just use that?

I'll have to double check the behavior when it's piped to a file.

pitaj commented:
I've removed the cargo-upgrades stuff and added the cargo-upgrade log to the PR and commit message instead.

It looks like pushing to a different repository is not at all simple. It requires a separate auth token and everything has to be done manually. It would significantly complicate things. Can you explain more about what your concern is regarding "noisiness"? As for secrets, I have to admit I'm not well acquainted with the security model there. Why do you think opening a PR from a branch in this repository is more risky?

Mark-Simulacrum commented on .github/workflows/dependencies.yml:
```diff
@@ -0,0 +1,115 @@
+          git add ./Cargo.lock
+          git commit --no-verify --file=commit.txt
+
+      - name: edit and reopen existing pull request
```
Is this addressing the case where the PR was already merged? Or where the PR was closed without merging? I think my current feeling is that we should not push commits or otherwise run updates until the previous PR merges. With at least current bors, reusing a single PR across possible r+'s is going to be more confusing and I think more likely to run into us not merging it in time before it gets bumped again by the automation here.

Can we bail out here if a PR is already open?

pitaj replied:
The behavior I had is this:

- If PR was closed without merging, reopen, edit, and push
- If PR was merged, push and open new PR
- If PR is already open, edit and push

My thinking was that someone will probably check in on the PR when they have time, and it shouldn't be a week outdated when they check it. However, I was operating under and assumption of the "Github merge button" and hadn't considered bors, so I'll have to change it to never reopen a closed PR.

So the new behavior will be:

- If PR is `S-waiting-on-bors`, skip
- If PR was closed for any reason, push and open new PR
- If PR is already open, push and edit

Mark-Simulacrum commented:
> It looks like pushing to a different repository is not at all simple. It requires a separate auth token and everything has to be done manually. It would significantly complicate things. Can you explain more about what your concern is regarding "noisiness"? As for secrets, I have to admit I'm not well acquainted with the security model there. Why do you think opening a PR from a branch in this repository is more risky?

When folks clone and/or git pull in this repo, any new branches that get touched are also pulled. I think I missed that this PR is always(?) updating a single branch, which doesn't seem that bad, I think that addresses that concern for me.

> As for secrets, I have to admit I'm not well acquainted with the security model there. Why do you think opening a PR from a branch in this repository is more risky?

Hm. So thinking more about this in practice I expect these PRs to get r+'d pretty quickly, so I don't think there's value in trying to go to a different repository. I think GitHub Actions exposes secrets configured in this repository to all branches in the repo, which it wouldn't do externally, but I think the risk/benefit here is pretty minimal.

Left a few inline comments but I think we can squash commits and move forward once the comments are resolved.

Mark-Simulacrum commented:
@rustbot author

pitaj commented:
> I think I missed that this PR is always(?) updating a single branch, which doesn't seem that bad, I think that addresses that concern for me.

Yes, it's always the `cargo_update` branch. And at most, that branch will have one commit relative to master with only one file changed, so it should have practically zero impact on clones or fetches.

pitaj commented:
Alright I've addressed those issues and rebased

@rustbot ready

Mark-Simulacrum commented on .github/workflows/dependencies.yml:
```diff
@@ -0,0 +1,137 @@
+          git commit --no-verify --file=commit.txt
+
+      - name: push
+        run: git push --no-verify -f -u origin cargo_update
```
I think this step needs to be moved after checking for existing PRs -- my thinking is that we should refuse to *update* a PR. We might be able to alter that to something like "no updates if labeled S-waiting-on-bors", but that might be too complex.

pitaj replied:
> I think this step needs to be moved after checking for existing PRs -- my thinking is that we should refuse to *update* a PR. We might be able to alter that to something like "no updates if labeled S-waiting-on-bors", but that might be too complex.

If you don't want to update a PR at all, that's fine. But if your concern is just bors:

- The whole thing is skipped when the existing PR is open and marked S-waiting-on-bors
- When the existing PR is closed, pushes don't update the existing PR, so we push and open a new PR
- Only when the existing PR is open and not S-waiting-on-bors, push and edit that PR

That said, moving the push after the edit or open would make this less reliant on that "pushing to a closed GitHub PR doesn't update it" behavior.

Mark-Simulacrum replied:
Ah, I missed that block earlier in the file -- happy to merge with or without that change. I think pushing to a closed PR makes it impossible to re-open it, so I think making this change is probably good, but non-critical. (We almost certainly will rarely or never be closing these PRs).

pitaj replied:
Okay wasn't able to get push-after-open to work, but if that's alright with you, we should be good to go.

pitaj commented:
Alright it looks like using cargo from bootstrap beta is [not gonna work as well as I'd hoped](https://github.com/pitaj/rust/actions/runs/5231612803/jobs/9445860224#step:4:23). I'll try switching back to nightly cargo.

Mark-Simulacrum commented:
@bors r+ rollup

Thanks! I think we can go ahead and land this, and try it out in practice. If we encounter issues can always back out and try again.

bors commented:
:pushpin: Commit 6d3ff105365318fe41c907c424c6ff2b2ba80dae has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 6d3ff105365318fe41c907c424c6ff2b2ba80dae -->
<!-- homu: {"type":"Approved","sha":"6d3ff105365318fe41c907c424c6ff2b2ba80dae","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

klensy commented:
I guess it's failed https://github.com/rust-lang/rust/pull/112746#issuecomment-1595903207?

pitaj commented:
> I guess it's failed https://github.com/rust-lang/rust/pull/112746#issuecomment-1595903207?

Yeah we're aware. Bot doesn't have enough privileges to push to that branch. I'm researching what exactly needs to be done to resolve that.

I'm not sure why it showed up in the PR, but we may want to look into excluding it from log analysis.
